scripts/publish_layers.sh

```diff
@@ -76,46 +92,50 @@ publish_layer() {
         --compatible-runtimes $aws_version_key \
                         | jq -r '.Version')
 
-    aws lambda add-layer-version-permission --layer-name $layer_name \
+    permission=$(aws lambda add-layer-version-permission --layer-name $layer_name \
         --version-number $version_nbr \
         --statement-id "release-$version_nbr" \
         --action lambda:GetLayerVersion --principal "*" \
-        --region $region
+        --region $region)
 
-    echo "Published layer for region $region, python version $aws_version_key, layer_name $layer_name, layer_version $version_nbr"
-}
-
-BATCH_SIZE=1
-PIDS=()
-
-wait_for_processes() {
-    for pid in "${PIDS[@]}"; do
-        wait $pid
-    done
-    PIDS=()
+    echo $version_nbr
 }
 
 for region in $REGIONS
 do
     echo "Starting publishing layer for region $region..."
-
     # Publish the layers for each version of python
-    i=0
-    for layer_name in "${LAYER_NAMES[@]}"; do
-        aws_version_key="${PYTHON_VERSIONS_FOR_AWS_CLI[$i]}"
-        layer_path="${LAYER_PATHS[$i]}"
-
-        publish_layer $region $layer_name $aws_version_key $layer_path &
-        PIDS+=($!)
-        if [ ${#PIDS[@]} -eq $BATCH_SIZE ]; then
-            wait_for_processes
+    for layer_name in "${LAYERS[@]}"; do
+        latest_version=$(aws lambda list-layer-versions --region $region --layer-name $layer_name --query 'LayerVersions[0].Version || `0`')
+        if [ $latest_version -ge $VERSION ]; then
+            echo "Layer $layer_name version $VERSION already exists in region $region, skipping..."
+            continue
+        elif [ $latest_version -lt $((VERSION-1)) ]; then
+            read -p "WARNING: The latest version of layer $layer_name in region $region is $latest_version, publish all the missing versions including $VERSION or EXIT the script (y/n)?" CONT
+            if [ "$CONT" != "y" ]; then
+                echo "Exiting"
+                exit 1
+            fi
         fi
 
-        i=$(expr $i + 1)
-
+        index=$(index_of_layer $layer_name)
+        aws_version_key="${PYTHON_VERSIONS_FOR_AWS_CLI[$index]}"
+        layer_path="${LAYER_PATHS[$index]}"
+
+        while [ $latest_version -lt $VERSION ]; do
+            latest_version=$(publish_layer $region $layer_name $aws_version_key $layer_path)
+            echo "Published version $latest_version for layer $layer_name in region $region"
+
+            # This shouldn't happen unless someone manually deleted the latest version, say 28
+            # and then try to republish it again. The published version is actually be 29, because
+            # Lambda layers are immutable and AWS will skip deleted version and use the next number. 
```

Comment:
👍  I guess if this ever does happen, the script can just be rerun with the higher layer version to catch the rest of the regions up.

Reply:
that's right